# UDBG_SNPs_rnd2
This project directory stores the analyses done on Jim Brosnan's UDBG data.


### Methods
Genotyping by Sequencing (GBS) through illumina sequencing and analysis by the UNEAK portion of the Tassel pipeline were prepared by Qi Sun out of the genomics group from Cornell University (Bradbury et al. 2007; Glaubitz et al. 2014). One of the limitations of the UNEAK pipeline is that its single nucleotide variant (SNV) calling algorithm relies on a diploid model. However, ultra dwarf bermuda grass sequenced in this project inlcuded diploid, triploid, and tetraploid cultivars. To this end an alternative approach was used to call SNVs. Sequence tags were extracted from the topm.bin libraries generated during the TagPairExport phase, using the UNEAK BinaryToTextPlugin. Using the tags generated from the UNEAK pipeline as a pseduo-reference, raw reads were mapped with bowtie2 v2.2.7 (Langmead, Ben, and Salzberg 2012). Processing of the alignment files was performed with the samtools v1.3 package including sorting, indexing, and merging (Li et al. 2009). The haplotype-based variants caller, Freebayes v1.0.2-15, was used to call SNVs for cultivars of varying ploidy including diploid, triploid, and tetraploid (Garrison and Marth 2012). Multidimensional scaling (MDS) plots were generated from these SNVs using PLINK v1.9, samples with less than one million raw reads were not included (Purcell et al. 2007). Results from calling SNVs using Freebayes for diploid, triploid, and tetraploid were combined and used as priors for re-running Freebayes for diploid, triploid, and tetraploid datasets. Python was used to filter each SNV, per individual, to a minimum read depth of five (van Rossum 1998). SNVs with less depth than five were not considered in further analysis. The most common genotypes, making up â‰¥60% of genotypes among individuals per cultivar, were assigned to each cultivar. Positions with SNVs between cultivars giving information to distinguish cultivars from each other were determined using Python. Alignment of all reads against the UNEAK pipeline generated tags was performed again along with all other following analyses after pooling reads of samples in each of the cultivars. Each offtype contained only one sample and were therefore unpooled. 
